src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -59,6 +59,17 @@ void GcsNodeManager::HandleRegisterNode(rpc::RegisterNodeRequest request,
     RAY_LOG(INFO) << "Finished registering node info, node id = " << node_id
                   << ", address = " << request.node_info().node_manager_address()
                   << ", node name = " << request.node_info().node_name();
+    if (request.node_info().is_head_node()) {
+      // mark the old head node as dead
+      auto head_node_it =
+          std::find_if(alive_nodes_.begin(), alive_nodes_.end(), [](const auto &node) {
+            return node.second->is_head_node();
+          });
```

Comment:
Since here we add the new head and then remove the old one. If we crash in between, we will have two heads after restarts so we need to mark all head nodes as dead.

Alternatively, we can mark the old one as dead before adding a new one. This guarantees that there can only be one alive head node in redis at any given point of time.

Reply:
we do this before add node in line 74 which adding the head node to alive_nodes.
can move this before `put_node` to table storage too

Reply:
What matters is the order in terms of redis.